scripts/ASTImportTest.sh

```diff
@@ -1,105 +1,174 @@
+SPLITSOURCES=$REPO_ROOT/scripts/splitSources.py
+
+# shellcheck source=scripts/common.sh
+source "${REPO_ROOT}/scripts/common.sh"
+
+function print_usage
+{
+  fail "Usage: ${0} ast [--exit-on-error]."
+}
+
+function print_used_commands
+{
+    local test_directory="$1"
+    local export_command="$2"
+    local import_command="$3"
+    printError "You can find the files used for this test here: ${test_directory}"
+    printError "Used commands for test:"
+    printError "# export"
+    echo "$ ${export_command}" >&2
+    printError "# import"
+    echo "$ ${import_command}" >&2
+}
+
+function print_stderr_stdout
+{
+    local error_message="$1"
+    local stderr_file="$2"
+    local stdout_file="$3"
+    printError "${error_message}"
+    printError ""
+    printError "stderr:"
+    cat "${stderr_file}" >&2
+    printError ""
+    printError "stdout:"
+    cat "${stdout_file}" >&2
+}
+
+IMPORT_TEST_TYPE=
+EXIT_ON_ERROR=0
+for PARAM in "$@"
+do
+  case "$PARAM" in
+    ast) IMPORT_TEST_TYPE="ast" ;;
+    --exit-on-error) EXIT_ON_ERROR=1 ;;
+    *) print_usage ;;
+  esac
+done
 
 SYNTAXTESTS_DIR="${REPO_ROOT}/test/libsolidity/syntaxTests"
 ASTJSONTESTS_DIR="${REPO_ROOT}/test/libsolidity/ASTJSON"
-NSOURCES="$(find "$SYNTAXTESTS_DIR" -type f | wc -l)"
-
-# DEV_DIR="${REPO_ROOT}/../tmp/contracts/"
-# NSOURCES="$(find $DEV_DIR -type f | wc -l)" #TODO use find command
 
 FAILED=0
 UNCOMPILABLE=0
 TESTED=0
 
-if [[ "$(find . -maxdepth 0 -type d -empty)" == "" ]]; then
-    echo "Test directory not empty. Skipping!"
-    exit 1
-fi
+function ast_import_export_equivalence
+{
+    local sol_file="$1"
+    local input_files=( "${@:2}" )
+
+    # export ast - save ast json as expected result (silently)
+    local export_command="$SOLC --combined-json ast --pretty-json --json-indent 4 ${input_files[*]} > expected.json 2> stderr_export.txt"
+    if ! $SOLC --combined-json ast --pretty-json --json-indent 4 "${input_files[@]}" > expected.json 2> stderr_export.txt
+    then
+        print_stderr_stdout "ERROR: AST export failed for input file $sol_file." ./stderr_export.txt ./expected.json
+        print_used_commands "$(pwd)" "${export_command}" "${import_command}"
+        return 1
+    fi
 
-# function tests whether exporting and importing again leaves the JSON ast unchanged
+    # (re)import ast - and export it again as obtained result (silently)
+    local import_command="$SOLC --import-ast --combined-json ast --pretty-json --json-indent 4 expected.json > obtained.json 2> stderr_import.txt"
+    if ! $SOLC --import-ast --combined-json ast --pretty-json --json-indent 4 expected.json > obtained.json 2> stderr_import.txt
+    then
+        print_stderr_stdout "ERROR: AST export failed for input file $sol_file." ./stderr_import.txt ./obtained.json
+        print_used_commands "$(pwd)" "${export_command}" "${import_command}"
+        return 1
+    fi
+
+    # compare expected and obtained ast's
+    if diff_files expected.json obtained.json
```

Comment:
maybe `diff_files` should return inverted return code? because `if diff_files [..]` reads also like "if there is a diff".. but actually there is exactly no difference detected e.g. `echo -n "✅"`.

Reply:
I think it's the most intiutive if it returns the same error code as plain `diff`. When I introduced `diff_values` it was really meant to be a replacement for `diff`, just one that does not require storing input in files.